### PR TITLE
SubMesh for Nedelec and Raviart-Thomas Bases

### PR DESCRIPTION
**Pull request by mlstowell**
This PR is studying the feasibility of adding support for Nedelec and Raviart-Thomas basis functions on SubMesh objects.

As a partial test of this new capability we add two new examples `ex34p.cpp` and `ex35p.cpp`. Example 34 solves a magnetostatics problem where the current density is computed on a domain derived `SubMesh`. Example 35 uses a port boundary condition to drive a wave equation in the full domain (this is a modified version of `ex22p`). The wave equation can use either H1, ND, or RT basis functions. The field used as the port boundary condition is computed as an eigenmode on a `SubMesh` object which is defined using a subset of the boundary attributes. Four new unit tests are also added.

This PR includes a significant redesign of the `DofTransformation` class although the resulting design should be backward compatible. The purpose of this redesign was to provide access to the underlying face transformations without tying them to a particular `FiniteElementSpace` since we need to transform degrees of freedom between a parent mesh and a sub-mesh which have distinct finite element spaces.

This PR also includes two new 2D meshes with non-trivial sets of boundary attributes which were used for testing. Although the test codes themselves do not appear as part of this PR I thought the meshes might be useful and/or interesting even without the tests (e.g. see `llnl-p3.mesh`).
llnl-p3

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#3379](https://github.com/mfem/mfem/pull/3379) | @mlstowell | @tzanio | @jamiebramwell + @psocratis + @jandrej | 6/4/23 | 7/28/23 | 7/29/23 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

**Comment by mlstowell** [comment]
The problem with invalid `ParSubMesh` boundaries in parallel seems to arise because the `faces_info` array does not contain information about elements from neighboring processors which abut local edges. I'm not yet sure why this info is not there nor how to obtain it.

**Comment by mlstowell** [comment]
The problem with high order Nedelec triangles is caused an old friend. Nedelec basis functions associated with the interior of element faces are oriented with the reference triangle or quadrilateral. However, neighboring elements sharing a common face, or in this case neighboring meshes sharing a common face, may choose different orientations for their reference elements.

This problem is resolved using data from the `DofTransformation` objects returned by `GetElementVDofs()` and similar methods. We will need to incorporate this `DofTransformation` data into the `ParTransferMap`.

One possibility is to build a `SparseMatrix` containing this data and add a MatVec to the transfer step. This is conceptually simple but in most cases this matrix will be the identity or differ from the identity only slightly. It may be more efficient to create a specialized operator which only touches the DoFs which need to be manipulated.

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

File: fem/doftrans.hpp [hunk 1]
```diff
@@ -15,19 +15,19 @@
...
 #include "intrules.hpp"
-#include "fe.hpp"
 
 namespace mfem
 {
 
-/** The DofTransformation class is an abstract base class for a family of
-    transformations that map local degrees of freedom (DoFs), contained within
-    individual elements, to global degrees of freedom, stored within
+/** The StatelessDofTransformation class is an abstract base class for a family
```
Comment (jandrej) [inline fem/doftrans.hpp:22]: Maybe describe what the term stateless means here.

File: mesh/submesh/ptransfermap.cpp [hunk 2]
```diff
@@ -41,9 +41,68 @@ ParTransferMap::ParTransferMap(const ParGridFunction &src,
...
+
+            const L2_FECollection *src_l2_fec =
+               dynamic_cast<const L2_FECollection*>(src_fec);
+            const L2_FECollection *dst_l2_fec =
+               dynamic_cast<const L2_FECollection*>(dst_fec);
+
+            if (src_l2_fec != NULL && dst_l2_fec != NULL)
+            {
+               // Source and destination are both lower dimension L2 spaces.
+               // Transfer them as the trace of an RT space if possible.
```
Comment (jandrej) [inline mesh/submesh/ptransfermap.cpp:69]: 🤯

**Review by jandrej — COMMENTED** [review]
I can mostly comment on the Mesh and SubMesh changes/additions. Those look good from my side, they fit into the interface nicely and provide features that can be used for other functionality as well.

Due to the changes in DofTransformation (an elemental part of everything that uses MFEM) I suggest adding another reviewer who is well versed in those parts of the code.

File: fem/doftrans.hpp (hunk 1 above)
Comment (mlstowell) [reply fem/doftrans.hpp:22]: Good call. I have expanded this description.

File: mesh/submesh/ptransfermap.cpp (hunk 2 above)
Comment (mlstowell) [reply mesh/submesh/ptransfermap.cpp:69]: I hope this is a positive reaction.

Comment (jandrej) [reply mesh/submesh/ptransfermap.cpp:69]: Yes.

**Review by jandrej — APPROVED** [review]

File: examples/ex35p.cpp [hunk 3]
```diff
@@ -0,0 +1,809 @@
+//                       MFEM Example 35 - Parallel Version
+//
+// Compile with: make ex35p
+//
+// Sample runs:  mpirun -np 4 ex35p -p 0 -o 2
```
[the remark this answers is not in the document]
Comment (mlstowell) [reply examples/ex35p.cpp:5]: Thanks, @psocratis! I struggled with these but I'm happy with how they turned out.

File: examples/ex34p.cpp [hunk 4]
```diff
@@ -0,0 +1,574 @@
+//                       MFEM Example 34 - Parallel Version
+//
+// Compile with: make ex34p
+//
+// Sample runs:  mpirun -np 4 ex34p -o 2
+//               mpirun -np 4 ex34p -o 2 -pa
```
[the remark this answers is not in the document]
Comment (mlstowell) [reply examples/ex34p.cpp:6]: I didn't think of that. I'll need to devise a modified problem to run on a mesh with tensor product elements... I think I can do that without too much effort.

File: examples/makefile [hunk 5]
```diff
@@ -183,3 +183,4 @@ clean-exec:
 	@rm -f ex23.mesh ex23-*.gf
 	@rm -f ex25.mesh ex25-*.gf ex25p-*.*
 	@rm -rf ex28_* ex28p_*
+	@rm -rf cond_mesh.* cond_j.* dsol.* port_mesh.* port_mode.*
```
[the remark this answers is not in the document]
Comment (mlstowell) [reply examples/makefile:186]: Done

File: examples/ex34p.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (mlstowell) [reply examples/ex34p.cpp:6]: I'm seeing an issue on processor boundaries of a hexahedral mesh. I didn't see any such issue on mixed meshes so the problem may be limited to shared quadrilateral faces within the `ParSubMesh`. I may need a couple days to get back to this and dig deeper.

[the remark this answers is not in the document]
Comment (mlstowell) [reply examples/ex34p.cpp]: I believe this is now fixed. I've also added support for a hex-only mesh in both examples so that we can test the partial assembly options.

I've very glad you caught this, @psocratis. Thanks!

**Review by psocratis — APPROVED** [review]

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Comment by tzanio** [comment]
Re-merged in `next` for testing...

**Comment by tzanio** [comment]
Re-merged in `next` for testing...

**Comment by v-dobrev** [comment]
I see that all three new examples support the `--device`/`-d` option, so I think we should add them to the list of the "device examples", here: https://github.com/mfem/mfem/blob/f97a770dcaa7498c93069ca7f27d25d9655acd86/examples/makefile#L30-L32 and here: https://github.com/mfem/mfem/blob/f97a770dcaa7498c93069ca7f27d25d9655acd86/examples/CMakeLists.txt#L120-L124

**Comment by tzanio** [comment]
@mlstowell, can you please merge `master` here and resolve conflicts?

Also, do you want to make the [above change](https://github.com/mfem/mfem/pull/3379#issuecomment-1648693462) suggested by @v-dobrev?

**Comment by tzanio** [comment]
Re-merged in `next` for testing...
